File: src/main/java/org/example/kalkulationsprogramm/controller/ZeitkontoKorrekturController.java
package org.example.kalkulationsprogramm.controller;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.ErfassungsQuelle;
import org.example.kalkulationsprogramm.domain.ZeitkontoKorrektur;
import org.example.kalkulationsprogramm.service.ZeitkontoKorrekturService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * REST Controller für Zeitkonto-Korrekturen.
 * Ermöglicht manuelle Ausgleichsbuchungen ohne Projektzuordnung.
 * 
 * GoBD-konform: Alle Änderungen werden unveränderlich protokolliert.
 */
@RestController
@RequestMapping("/api/zeitkonto/korrekturen")
@RequiredArgsConstructor
public class ZeitkontoKorrekturController {

    private final ZeitkontoKorrekturService korrekturService;

    /**
     * Erstellt eine neue Zeitkonto-Korrektur.
     * 
     * Body: {
     * "mitarbeiterId": 1,
     * "stunden": 8.5, // Positiv = Gutschrift, Negativ = Abzug
     * "datum": "2025-12-21",
     * "grund": "Überstundenausgleich Q4 2025",
     * "erstelltVonId": 1,
     * "typ": "STUNDEN" // oder "URLAUB"
     * }
     */
    @PostMapping
    public ResponseEntity<?> erstelleKorrektur(@RequestBody Map<String, Object> body) {
        try {
            Long mitarbeiterId = getLong(body, "mitarbeiterId");
            BigDecimal stunden = getBigDecimal(body, "stunden");
            LocalDate datum = LocalDate.parse((String) body.get("datum"));
            String grund = (String) body.get("grund");
            Long erstelltVonId = getLong(body, "erstelltVonId");

            // Typ parsen (optional, default STUNDEN)
            String typStr = (String) body.get("typ");
            org.example.kalkulationsprogramm.domain.KorrekturTyp typ = typStr != null
                    ? org.example.kalkulationsprogramm.domain.KorrekturTyp.valueOf(typStr)
                    : org.example.kalkulationsprogramm.domain.KorrekturTyp.STUNDEN;

            ZeitkontoKorrektur korrektur = korrekturService.erstelleKorrektur(
                    mitarbeiterId, stunden, datum, grund, erstelltVonId, ErfassungsQuelle.DESKTOP, typ);

            return ResponseEntity.status(HttpStatus.CREATED).body(korrekturToMap(korrektur));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("error", "Fehler beim Erstellen: " + e.getMessage()));
        }
    }

    /**
     * Aktualisiert eine bestehende Korrektur (GoBD-konform mit Audit).
     * 
     * Body: {
     * "stunden": 10.0,
     * "grund": "Korrigierter Wert",
     * "bearbeiterId": 1,
     * "aenderungsgrund": "Tippfehler korrigiert"
     * }
     */
    @PutMapping("/{id}")
    public ResponseEntity<?> aendereKorrektur(@PathVariable Long id, @RequestBody Map<String, Object> body) {
        try {
            BigDecimal stunden = body.get("stunden") != null ? getBigDecimal(body, "stunden") : null;
            String grund = (String) body.get("grund");
            Long bearbeiterId = getLong(body, "bearbeiterId");
            String aenderungsgrund = (String) body.get("aenderungsgrund");

            ZeitkontoKorrektur korrektur = korrekturService.aendereKorrektur(
                    id, stunden, grund, bearbeiterId, aenderungsgrund, ErfassungsQuelle.DESKTOP);

            return ResponseEntity.ok(korrekturToMap(korrektur));
        } catch (IllegalArgumentException | IllegalStateException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    /**
     * Storniert eine Korrektur (GoBD: keine physische Löschung!).
     * 
     * Body: {
     * "bearbeiterId": 1,
     * "stornierungsgrund": "Fehlerhafte Buchung"
     * }
     */
    @DeleteMapping("/{id}")
    public ResponseEntity<?> storniereKorrektur(@PathVariable Long id, @RequestBody Map<String, Object> body) {
        try {
            Long bearbeiterId = getLong(body, "bearbeiterId");
            String stornierungsgrund = (String) body.get("stornierungsgrund");

            korrekturService.storniereKorrektur(id, bearbeiterId, stornierungsgrund, ErfassungsQuelle.DESKTOP);

            return ResponseEntity.ok(Map.of("success", true, "message", "Korrektur wurde storniert"));
        } catch (IllegalArgumentException | IllegalStateException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        }
    }

    /**
     * Gibt alle aktiven Korrekturen eines Mitarbeiters zurück.
     */
    @GetMapping("/mitarbeiter/{mitarbeiterId}")
    public ResponseEntity<List<Map<String, Object>>> getKorrekturenByMitarbeiter(
            @PathVariable Long mitarbeiterId,
            @RequestParam(defaultValue = "false") boolean alleAnzeigen) {

        List<ZeitkontoKorrektur> korrekturen = alleAnzeigen
                ? korrekturService.getAlleKorrekturenByMitarbeiter(mitarbeiterId)
                : korrekturService.getAktiveKorrekturenByMitarbeiter(mitarbeiterId);

        List<Map<String, Object>> result = korrekturen.stream()
                .map(this::korrekturToMap)
                .collect(Collectors.toList());

        return ResponseEntity.ok(result);
    }

    /**
     * Gibt die Audit-Historie einer Korrektur zurück (für GoBD-Prüfung).
     */
    @GetMapping("/{id}/historie")
    public ResponseEntity<List<Map<String, Object>>> getAuditHistorie(@PathVariable Long id) {
        return ResponseEntity.ok(korrekturService.getAuditHistorie(id));
    }

    /**
     * Gibt die Summe aller aktiven Korrekturen eines Mitarbeiters für ein Jahr
     * zurück.
     */
    @GetMapping("/mitarbeiter/{mitarbeiterId}/summe")
    public ResponseEntity<Map<String, Object>> getSumme(
            @PathVariable Long mitarbeiterId,
            @RequestParam(defaultValue = "0") int jahr) {

        if (jahr == 0) {
            jahr = LocalDate.now().getYear();
        }

        BigDecimal summe = korrekturService.summiereAktiveKorrekturen(mitarbeiterId, jahr);

        return ResponseEntity.ok(Map.of(
                "mitarbeiterId", mitarbeiterId,
                "jahr", jahr,
                "summeStunden", summe));
    }

    // ==================== Helper Methods ====================

    private Map<String, Object> korrekturToMap(ZeitkontoKorrektur k) {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("id", k.getId());
        map.put("mitarbeiterId", k.getMitarbeiter().getId());
        map.put("mitarbeiterName", k.getMitarbeiter().getVorname() + " " + k.getMitarbeiter().getNachname());
        map.put("datum", k.getDatum().toString());
        map.put("stunden", k.getStunden());
        map.put("grund", k.getGrund());
        map.put("version", k.getVersion());
        map.put("typ", k.getTyp().name());
        map.put("erstelltAm", k.getErstelltAm().toString());
        map.put("erstelltVon", k.getErstelltVon() != null
                ? k.getErstelltVon().getVorname() + " " + k.getErstelltVon().getNachname()
                : null);
        map.put("storniert", k.getStorniert());
        if (k.getStorniert()) {
            map.put("storniertAm", k.getStorniertAm() != null ? k.getStorniertAm().toString() : null);
            map.put("storniertVon", k.getStorniertVon() != null
                    ? k.getStorniertVon().getVorname() + " " + k.getStorniertVon().getNachname()
                    : null);
            map.put("stornierungsgrund", k.getStornierungsgrund());
        }
        return map;
    }

    private Long getLong(Map<String, Object> body, String key) {
        Object value = body.get(key);
        if (value == null) {
            throw new IllegalArgumentException(key + " ist ein Pflichtfeld");
        }
        return ((Number) value).longValue();
    }

    private BigDecimal getBigDecimal(Map<String, Object> body, String key) {
        Object value = body.get(key);
        if (value == null) {
            throw new IllegalArgumentException(key + " ist ein Pflichtfeld");
        }
        if (value instanceof BigDecimal decimal) {
            return decimal;
        }
        return new BigDecimal(value.toString());
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/ZeitkontoKorrekturController.kt
package org.example.kalkulationsprogramm.controller

import org.example.kalkulationsprogramm.domain.ErfassungsQuelle
import org.example.kalkulationsprogramm.domain.KorrekturTyp
import org.example.kalkulationsprogramm.domain.ZeitkontoKorrektur
import org.example.kalkulationsprogramm.service.ZeitkontoKorrekturService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import java.math.BigDecimal
import java.time.LocalDate
import java.util.LinkedHashMap

@RestController
@RequestMapping("/api/zeitkonto/korrekturen")
class ZeitkontoKorrekturController(
    private val korrekturService: ZeitkontoKorrekturService,
) {
    @PostMapping
    fun erstelleKorrektur(@RequestBody body: Map<String, Any?>): ResponseEntity<*> =
        try {
            val mitarbeiterId = getLong(body, "mitarbeiterId")
            val stunden = getBigDecimal(body, "stunden")
            val datum = LocalDate.parse(body["datum"] as String)
            val grund = body["grund"] as String?
            val erstelltVonId = getLong(body, "erstelltVonId")
            val typ = (body["typ"] as String?)?.let(KorrekturTyp::valueOf) ?: KorrekturTyp.STUNDEN

            val korrektur = korrekturService.erstelleKorrektur(
                mitarbeiterId,
                stunden,
                datum,
                grund,
                erstelltVonId,
                ErfassungsQuelle.DESKTOP,
                typ,
            )
            ResponseEntity.status(HttpStatus.CREATED).body(korrekturToMap(korrektur))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to ex.message))
        } catch (ex: Exception) {
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(mapOf("error" to "Fehler beim Erstellen: ${ex.message}"))
        }

    @PutMapping("/{id}")
    fun aendereKorrektur(
        @PathVariable id: Long,
        @RequestBody body: Map<String, Any?>,
    ): ResponseEntity<*> =
        try {
            val stunden = if (body["stunden"] != null) getBigDecimal(body, "stunden") else null
            val grund = body["grund"] as String?
            val bearbeiterId = getLong(body, "bearbeiterId")
            val aenderungsgrund = body["aenderungsgrund"] as String?

            val korrektur = korrekturService.aendereKorrektur(
                id,
                stunden,
                grund,
                bearbeiterId,
                aenderungsgrund,
                ErfassungsQuelle.DESKTOP,
            )
            ResponseEntity.ok(korrekturToMap(korrektur))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to ex.message))
        } catch (ex: IllegalStateException) {
            ResponseEntity.badRequest().body(mapOf("error" to ex.message))
        }

    @DeleteMapping("/{id}")
    fun storniereKorrektur(
        @PathVariable id: Long,
        @RequestBody body: Map<String, Any?>,
    ): ResponseEntity<*> =
        try {
            val bearbeiterId = getLong(body, "bearbeiterId")
            val stornierungsgrund = body["stornierungsgrund"] as String?
            korrekturService.storniereKorrektur(id, bearbeiterId, stornierungsgrund, ErfassungsQuelle.DESKTOP)
            ResponseEntity.ok(mapOf("success" to true, "message" to "Korrektur wurde storniert"))
        } catch (ex: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to ex.message))
        } catch (ex: IllegalStateException) {
            ResponseEntity.badRequest().body(mapOf("error" to ex.message))
        }

    @GetMapping("/mitarbeiter/{mitarbeiterId}")
    fun getKorrekturenByMitarbeiter(
        @PathVariable mitarbeiterId: Long,
        @RequestParam(defaultValue = "false") alleAnzeigen: Boolean,
    ): ResponseEntity<List<Map<String, Any?>>> {
        val korrekturen = if (alleAnzeigen) {
            korrekturService.getAlleKorrekturenByMitarbeiter(mitarbeiterId)
        } else {
            korrekturService.getAktiveKorrekturenByMitarbeiter(mitarbeiterId)
        }
        return ResponseEntity.ok(korrekturen.map(::korrekturToMap))
    }

    @GetMapping("/{id}/historie")
    fun getAuditHistorie(@PathVariable id: Long): ResponseEntity<List<Map<String, Any?>>> =
        ResponseEntity.ok(korrekturService.getAuditHistorie(id))

    @GetMapping("/mitarbeiter/{mitarbeiterId}/summe")
    fun getSumme(
        @PathVariable mitarbeiterId: Long,
        @RequestParam(defaultValue = "0") jahr: Int,
    ): ResponseEntity<Map<String, Any>> {
        val effectiveJahr = if (jahr == 0) LocalDate.now().year else jahr
        val summe = korrekturService.summiereAktiveKorrekturen(mitarbeiterId, effectiveJahr)
        return ResponseEntity.ok(
            mapOf(
                "mitarbeiterId" to mitarbeiterId,
                "jahr" to effectiveJahr,
                "summeStunden" to summe,
            ),
        )
    }

    private fun korrekturToMap(korrektur: ZeitkontoKorrektur): Map<String, Any?> {
        val map = LinkedHashMap<String, Any?>()
        val mitarbeiter = korrektur.mitarbeiter
        map["id"] = korrektur.id
        map["mitarbeiterId"] = mitarbeiter?.id
        map["mitarbeiterName"] = "${mitarbeiter?.vorname} ${mitarbeiter?.nachname}"
        map["datum"] = korrektur.datum?.toString()
        map["stunden"] = korrektur.stunden
        map["grund"] = korrektur.grund
        map["version"] = korrektur.version
        map["typ"] = korrektur.typ?.name
        map["erstelltAm"] = korrektur.erstelltAm?.toString()
        map["erstelltVon"] = korrektur.erstelltVon?.let { "${it.vorname} ${it.nachname}" }
        map["storniert"] = korrektur.storniert
        if (korrektur.storniert == true) {
            map["storniertAm"] = korrektur.storniertAm?.toString()
            map["storniertVon"] = korrektur.storniertVon?.let { "${it.vorname} ${it.nachname}" }
            map["stornierungsgrund"] = korrektur.stornierungsgrund
        }
        return map
    }

    private fun getLong(body: Map<String, Any?>, key: String): Long {
        val value = body[key] ?: throw IllegalArgumentException("$key ist ein Pflichtfeld")
        return (value as Number).toLong()
    }

    private fun getBigDecimal(body: Map<String, Any?>, key: String): BigDecimal {
        val value = body[key] ?: throw IllegalArgumentException("$key ist ein Pflichtfeld")
        return if (value is BigDecimal) value else BigDecimal(value.toString())
    }
}
